### PR TITLE
Implement OAuth authentication for libsplinter

### PR DESCRIPTION
**Pull request by ltseeley**
Adds OAuth endpoints to the Splinter REST API

This can be tested by checking out https://github.com/ltseeley/splinter/tree/rest-api-oauth-impl-test, which contains two extra commits for testing out authenticating with GitHub and Google.

### CODE REVIEW

File: libsplinter/src/auth/oauth/error.rs [hunk 1]
```diff
@@ -17,6 +17,28 @@
 use std::error::Error;
 use std::fmt;
 
+/// An unrecoverable error that can occur when using an OAuth client
+#[derive(Debug)]
+pub struct OAuthClientError(pub String);
```
Comment (vaporos) [inline libsplinter/src/auth/oauth/error.rs:22]: This form isn't consistent with our recent discussions that stick this in a "message" field.

File: libsplinter/src/auth/oauth/rest_api/actix/login.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+use actix_web::{http::header::LOCATION, HttpResponse};
+use futures::future::IntoFuture;
+
+use crate::auth::oauth::OAuthClient;
+use crate::protocol;
+use crate::rest_api::{ErrorResponse, Method, ProtocolVersionRangeGuard, Resource};
+
+pub fn make_login_route(client: OAuthClient) -> Resource {
+    Resource::build("/oauth/login")
```
Comment (vaporos) [inline libsplinter/src/auth/oauth/rest_api/actix/login.rs:25]: What's the expected flow?

1. Make a request on an endpoint.
2. Get auth denied.
3. GET /oauth/login
4. ...

or something else?

File: libsplinter/src/auth/oauth/rest_api/actix/login.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+//! The `GET /oauth/login` endpoint for getting the authorization request URL for the provider.
+
+use actix_web::{http::header::LOCATION, HttpResponse};
+use futures::future::IntoFuture;
+
+use crate::auth::oauth::OAuthClient;
+use crate::protocol;
+use crate::rest_api::{ErrorResponse, Method, ProtocolVersionRangeGuard, Resource};
+
+pub fn make_login_route(client: OAuthClient) -> Resource {
```
Comment (vaporos) [inline libsplinter/src/auth/oauth/rest_api/actix/login.rs:24]: Calling this route /login is a bit confusing in that its pretty common to POST to a /login route for username/password auth, where after the operation you are logged in.

Comment (ltseeley) [reply libsplinter/src/auth/oauth/rest_api/actix/login.rs:24]: Would `/oauth/authorize` be better?

File: libsplinter/src/auth/oauth/rest_api/actix/login.rs (hunk 2 above)
Comment (ltseeley) [reply libsplinter/src/auth/oauth/rest_api/actix/login.rs:25]: Yes, that's the expected flow.

File: libsplinter/src/auth/oauth/error.rs (hunk 1 above)
Comment (ltseeley) [reply libsplinter/src/auth/oauth/error.rs]: Updated to use a `message` field.

File: libsplinter/src/auth/oauth/rest_api/actix/login.rs (hunk 3 above)
Comment (vaporos) [reply libsplinter/src/auth/oauth/rest_api/actix/login.rs:24]: Almost feels more like retrieving config -- /oauth/config or something. Maybe we work it out in a subsequent PR.

**Review by vaporos — APPROVED** [review]

**Review by peterschwarz — APPROVED** [review]

**Pull request merged**